src/Tools/Replay/Replay.cs

```diff
@@ -0,0 +1,273 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Basic.CompilerLog.Util;
+using Microsoft.CodeAnalysis.CommandLine;
+using Microsoft.CodeAnalysis.Options;
+using Mono.Options;
+
+var options = ParseOptions(args);
+if (Directory.Exists(options.OutputDirectory))
+{
+    Directory.Delete(options.OutputDirectory, recursive: true);
+}
+Directory.CreateDirectory(options.OutputDirectory);
+Directory.CreateDirectory(options.TempDirectory);
+
+return await RunAsync(options).ConfigureAwait(false);
+
+static ReplayOptions ParseOptions(string[] args)
+{
+    string? outputDirectory = null;
+    string? binlogPath = null;
+    int maxParallel = 6;
+
+    var options = new Mono.Options.OptionSet()
+    {
+        { "b|binlogPath=", "The binary log to relpay", v => binlogPath = v },
+        { "o|outputDirectory=", "The directory to output the build results", v => outputDirectory = v },
+        { "m|maxParallel=", "The maximum number of parallel builds", (int v) => maxParallel = v },
+    };
+
+    var rest = options.Parse(args);
+    if (rest.Count == 1)
+    {
+        binlogPath = rest[0];
+    }
+    else if (rest.Count > 1)
+    {
+        throw new Exception("Too many arguments");
+    }
+
+    if (string.IsNullOrEmpty(binlogPath))
+    {
+        throw new Exception("Missing binlogPath");
+    }
+
+    if (string.IsNullOrEmpty(outputDirectory))
+    {
+        outputDirectory = Path.Combine(Environment.CurrentDirectory, "output");
+    }
+
+    return new ReplayOptions(
+        pipeName: Guid.NewGuid().ToString(),
+        clientDirectory: AppDomain.CurrentDomain.BaseDirectory!,
+        outputDirectory: outputDirectory,
+        binlogPath: binlogPath,
+        maxParallel: maxParallel);
+}
+
+static async Task<int> RunAsync(ReplayOptions options)
+{
+    Console.WriteLine($"Binary Log: {options.BinlogPath}");
+    Console.WriteLine($"Client Directory: {options.ClientDirectory}");
+    Console.WriteLine($"Output Directory: {options.OutputDirectory}");
+    Console.WriteLine($"Pipe Name: {options.PipeName}");
+    Console.WriteLine($"Parallel: {options.MaxParallel}");
+    Console.WriteLine();
+    Console.WriteLine("Starting server");
+
+    using var compilerServerLogger = new CompilerServerLogger("replay", Path.Combine(options.OutputDirectory, "server.log"));
+    if (!BuildServerConnection.TryCreateServer(options.ClientDirectory, options.PipeName, compilerServerLogger))
+    {
+        Console.WriteLine("Failed to start server");
+        return 1;
+    }
+
+    try
+    {
+        var stopwatch = new Stopwatch();
+        stopwatch.Start();
+
+        var compilerCalls = ReadAllCompilerCalls(options.BinlogPath);
```

Comment:
Should reading compiler calls be outside the stopwatch?

Reply:
Good catch. Yeah that call should be fast compared to compilation times but the point is to measure compilation time so should move this out.